File: python_pipeline/src/python_pipeline/models/ticker.py
"""
Ticker data model for CrypticsLabBot.
Implements Pydantic models for ticker data.
"""

import datetime
import random
from typing import ClassVar

from pydantic import Field

from python_pipeline.models.model_base import ModelBase


class ThalexTicker(ModelBase):
    """Ticker model based on the exact Thalex WebSocket data structure"""
    
    # Class variables for the ModelBase
    model_name: ClassVar[str] = "Ticker"
    
    # Fields matching Thalex data format but with Avro compatibility
    instrument_name: str = "BTC-PERPETUAL"
    mark_price: float = Field(..., description="Current mark price")
    mark_timestamp: float = Field(..., description="Timestamp of mark price")
    best_bid_price: float = Field(..., description="Best bid price")
    best_bid_amount: float = Field(..., description="Best bid amount")
    best_ask_price: float = Field(..., description="Best ask price")
    best_ask_amount: float = Field(..., description="Best ask amount")
    last_price: float = Field(..., description="Last traded price")
    delta: float = Field(..., description="Delta value")
    volume_24h: float = Field(..., description="24-hour trading volume")
    value_24h: float = Field(..., description="24-hour trading value")
    low_price_24h: float = Field(..., description="24-hour low price")
    high_price_24h: float = Field(..., description="24-hour high price")
    change_24h: float = Field(..., description="24-hour price change")
    index_price: float = Field(..., description="Index price")
    forward: float = Field(..., description="Forward price")
    funding_mark: float = Field(..., description="Funding mark")
    funding_rate: float = Field(..., description="Funding rate")
    collar_low: float = Field(..., description="Lower collar limit")
    collar_high: float = Field(..., description="Upper collar limit")
    realised_funding_24h: float = Field(..., description="24-hour realized funding")
    average_funding_rate_24h: float = Field(..., description="24-hour average funding rate")
    open_interest: float = Field(..., description="Open interest")

    @classmethod
    def generate(cls) -> 'ThalexTicker':
        """Generate sample ticker data based on actual Thalex data"""
        base_price = random.uniform(95000, 97000)
        bid_price = base_price - random.uniform(10, 30)
        ask_price = base_price + random.uniform(10, 30)
        
        return cls(
            mark_price=base_price,
            mark_timestamp=datetime.datetime.now().timestamp(),
            best_bid_price=bid_price,
            best_bid_amount=random.uniform(0.01, 0.1),
            best_ask_price=ask_price,
            best_ask_amount=random.uniform(0.01, 0.1),
            last_price=base_price + random.uniform(-50, 50),
            delta=1.0,
            volume_24h=random.uniform(700, 800),
            value_24h=random.uniform(68000000, 69000000),
            low_price_24h=random.uniform(93000, 94000),
            high_price_24h=random.uniform(97000, 98000),
            change_24h=random.uniform(2000, 3500),
            index_price=base_price - random.uniform(-10, 10),
            forward=base_price - random.uniform(-10, 10),
            funding_mark=random.uniform(0, 0.001),
            funding_rate=random.uniform(0, 0.001),
            collar_low=base_price * 0.99,
            collar_high=base_price * 1.01,
            realised_funding_24h=random.uniform(0, 0.001),
            average_funding_rate_24h=random.uniform(0, 0.001),
            open_interest=random.uniform(3700, 3800)
        )

File: rust_tradingengine/src/domain/model/ticker.rs
use anyhow::{anyhow, Result};
use serde::{Serialize, Deserialize};
use serde_json::Value;

/// Ticker data structure to represent market data across the application
/// This is the single, consolidated Ticker model for the entire application
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Ticker {
    pub instrument_name: String,
    pub mark_price: f64,
    pub mark_timestamp: f64,
    pub best_bid_price: f64,
    pub best_bid_amount: f64,
    pub best_ask_price: f64,
    pub best_ask_amount: f64,
    pub last_price: f64,
    pub delta: f64,
    pub volume_24h: f64,
    pub value_24h: f64,
    pub low_price_24h: f64,
    pub high_price_24h: f64,
    pub change_24h: f64,
    pub index_price: f64,
    pub forward: f64, 
    pub funding_mark: f64,
    pub funding_rate: f64,
    pub collar_low: f64,
    pub collar_high: f64,
    pub realised_funding_24h: f64,
    pub average_funding_rate_24h: f64,
    pub open_interest: f64,
    pub processing_timestamp: Option<f64>,
}

impl Ticker {
    /// Create a new Ticker with default values
    pub fn new(instrument_name: String) -> Self {
        Self {
            instrument_name,
            mark_price: 0.0,
            mark_timestamp: 0.0,
            best_bid_price: 0.0,
            best_bid_amount: 0.0,
            best_ask_price: 0.0,
            best_ask_amount: 0.0,
            last_price: 0.0,
            delta: 0.0,
            volume_24h: 0.0,
            value_24h: 0.0,
            low_price_24h: 0.0,
            high_price_24h: 0.0,
            change_24h: 0.0,
            index_price: 0.0,
            forward: 0.0,
            funding_mark: 0.0,
            funding_rate: 0.0,
            collar_low: 0.0,
            collar_high: 0.0,
            realised_funding_24h: 0.0,
            average_funding_rate_24h: 0.0,
            open_interest: 0.0,
            processing_timestamp: None,
        }
    }

    /// Create a Ticker from JSON notification data
    pub fn from_json(data: &Value, instrument_name: String) -> Result<Self> {
        // Get current time as processing timestamp
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs_f64();
            
        Ok(Self {
            instrument_name,
            mark_price: data["mark_price"].as_f64().ok_or_else(|| anyhow!("Missing mark_price"))?,
            mark_timestamp: data["mark_timestamp"].as_f64().ok_or_else(|| anyhow!("Missing mark_timestamp"))?,
            best_bid_price: data.get("best_bid_price").and_then(|v| v.as_f64()).unwrap_or(0.0),
            best_bid_amount: data.get("best_bid_amount").and_then(|v| v.as_f64()).unwrap_or(0.0),
            best_ask_price: data.get("best_ask_price").and_then(|v| v.as_f64()).unwrap_or(0.0),
            best_ask_amount: data.get("best_ask_amount").and_then(|v| v.as_f64()).unwrap_or(0.0),
            last_price: data.get("last_price").and_then(|v| v.as_f64()).unwrap_or(0.0),
            delta: data.get("delta").and_then(|v| v.as_f64()).unwrap_or(0.0),
            volume_24h: data.get("volume_24h").and_then(|v| v.as_f64()).unwrap_or(0.0),
            value_24h: data.get("value_24h").and_then(|v| v.as_f64()).unwrap_or(0.0),
            low_price_24h: data.get("low_price_24h").and_then(|v| v.as_f64()).unwrap_or(0.0),
            high_price_24h: data.get("high_price_24h").and_then(|v| v.as_f64()).unwrap_or(0.0),
            change_24h: data.get("change_24h").and_then(|v| v.as_f64()).unwrap_or(0.0),
            index_price: data["index"].as_f64().ok_or_else(|| anyhow!("Missing index"))?,
            forward: data.get("forward").and_then(|v| v.as_f64()).unwrap_or(0.0),
            funding_mark: data.get("funding_mark").and_then(|v| v.as_f64()).unwrap_or(0.0),
            funding_rate: data["funding_rate"].as_f64().ok_or_else(|| anyhow!("Missing funding_rate"))?,
            collar_low: data.get("collar_low").and_then(|v| v.as_f64()).unwrap_or(0.0),
            collar_high: data.get("collar_high").and_then(|v| v.as_f64()).unwrap_or(0.0),
            realised_funding_24h: data.get("realised_funding_24h").and_then(|v| v.as_f64()).unwrap_or(0.0),
            average_funding_rate_24h: data.get("average_funding_rate_24h").and_then(|v| v.as_f64()).unwrap_or(0.0),
            open_interest: data.get("open_interest").and_then(|v| v.as_f64()).unwrap_or(0.0),
            processing_timestamp: Some(now),
        })
    }

    /// Get best bid price, returning None if not available
    pub fn best_bid(&self) -> Option<f64> {
        if self.best_bid_price > 0.0 {
            Some(self.best_bid_price)
        } else {
            None
        }
    }

    /// Get best ask price, returning None if not available
    pub fn best_ask(&self) -> Option<f64> {
        if self.best_ask_price > 0.0 {
            Some(self.best_ask_price)
        } else {
            None
        }
    }
}
